tensorboard/webapp/app_routing/effects/app_routing_effects.ts

```diff
@@ -47,29 +47,62 @@ import {
 } from '../internal_utils';
 import {Location} from '../location';
 import {ProgrammaticalNavigationModule} from '../programmatical_navigation_module';
-import {RouteConfigs} from '../route_config';
+import {RouteConfigs, RouteMatch} from '../route_config';
 import {RouteRegistryModule} from '../route_registry_module';
 import {
   getActiveRoute,
   getActiveNamespaceId,
 } from '../store/app_routing_selectors';
-import {Navigation, Route, RouteKind, RouteParams} from '../types';
+import {Route, RouteKind, RouteParams} from '../types';
 
 /** @typehack */ import * as _typeHackNgrxEffects from '@ngrx/effects';
 /** @typehack */ import * as _typeHackModels from '@ngrx/store/src/models';
 /** @typehack */ import * as _typeHackStore from '@ngrx/store';
 
 const initAction = createAction('[App Routing] Effects Init');
 
-interface InternalNavigation extends Navigation {
-  browserInitiated?: boolean;
+// Note: The progression of transformation of navigation data is:
+// InternalNavigation -> InternalRouteMatch -> InternalRoute
+
+interface InternalNavigation {
+  pathname: string;
+  options: NavigationOptions;
+}
+
+interface InternalRouteMatch {
+  routeMatch: RouteMatch;
+  options: NavigationOptions;
 }
 
-interface NavigationOptions {
-  replaceState: boolean;
-  resetNamespacedState: boolean;
+interface InternalRoute {
+  route: Route;
+  options: NavigationOptions;
 }
 
+enum NamespaceOption {
+  // No change in namespace has been requested.
+  UNCHANGED,
```

Comment:
just a thought, if we rename `NamespaceOption` to `NamespaceUpdateOption`
then all the item naming makes sense to me.

Reply:
Done.